src/Hamiltonians/correlation_functions.jl

```diff
@@ -236,3 +236,153 @@ function get_offdiagonal(
+    address, value = excitation(add, (dst,), (src,))
+    return address, value / num_modes(add)
+end
+
+function diagonal_element(::SuperfluidCorrelator{0}, add::SingleComponentFockAddress)
+    return num_particles(add) / num_modes(add)
+end
+function diagonal_element(::SuperfluidCorrelator{D}, add::SingleComponentFockAddress) where {D}
+    return 0.0
+end
+
+
+"""
+    StringCorrelator(d::Int) <: AbstractHamiltonian{Float64}
+
+Two-body operator for string correlation between sites separated by `d` with `0 ≤ d < M`
+
+```math
+    \\hat{C}_{\\text{string}}(d) = \\frac{1}{M} \\sum_{j}^{M} \\delta n_j (e^{i \\pi \\sum_{j \\leq k < j + d} \\delta n_k}) \\delta n_{j+d}
+```
+Here ``\\bar{n} = N/M`` is the mean filling number of lattice sites with ``N`` and ``M`` being the number of particles and number of lattice sites respectively. ``\\delta \\hat{n}_j = \\hat{n}_j - \\bar{n}`` is the boson number deviation from the mean filling number.
+
+Assumes a one-dimensional lattice with periodic boundary conditions. 
+
+# Arguments
+- `d::Integer`: distance between sites. 
+
+# See also 
+
+* [`HubbardReal1D`](@ref)
+* [`G2RealCorrelator`](@ref)
+* [`SuperfluidCorrelator`](@ref)
+* [`AbstractHamiltonian`](@ref)
+* [`AllOverlaps`](@ref)
+
+"""
+struct StringCorrelator{D} <: AbstractHamiltonian{Float64}
+end
+
+StringCorrelator(d::Int) = StringCorrelator{d}()
+
+function Base.show(io::IO, ::StringCorrelator{D}) where {D}
+    print(io, "StringCorrelator($D)")
+end
+
+LOStructure(::Type{<:StringCorrelator}) = IsDiagonal()
+
+function diagonal_element(::StringCorrelator{0}, add::SingleComponentFockAddress)
+    M = num_modes(add)
+    N = num_particles(add)
+    n̄ = N/M
+    v = onr(add)
+
+    result = 0.0
+    for i in eachindex(v)
+        result += (v[i] - n̄)^2
+    end
+
+    return result / M
+end
+
+num_offdiagonals(::StringCorrelator, ::SingleComponentFockAddress) = 0
+
+function diagonal_element(::StringCorrelator{D}, add::SingleComponentFockAddress) where {D}
+    M = num_modes(add)
+    N = num_particles(add)
+    d = mod(D, M)
+
+    if iszero(N % M)
+        return _string_diagonal_real(d, add)
+    else
+        return _string_diagonal_complex(d, add)
+    end
+end
+
+function _string_diagonal_complex(d, add)
+    M = num_modes(add)
+    N = num_particles(add)
+    n̄ = N/M
+    v = onr(add)
+
+    result = ComplexF64(0)
+    for i in eachindex(v)
+        phase_sum = sum((v[mod1(k, M)] - n̄) for k in i:1:(i+d-1))
+
+        result += (v[i] - n̄) * exp(pi * im * phase_sum) * (v[mod1(i + d, M)] - n̄)
+    end
+
+    return result / M
+end
+function _string_diagonal_real(d, add)
+    M = num_modes(add)
+    N = num_particles(add)
+    n̄ = N ÷ M
+    v = onr(add)
+
+    result = 0.0
+    for i in eachindex(v)
+        phase_sum = sum((v[mod1(k, M)] - n̄) for k in i:1:(i+d-1))
+
+        result += (v[i] - n̄) * (-1)^phase_sum * (v[mod1(i + d, M)] - n̄)
+    end
+
+    return result / M
+end
```

Comment:
```suggestion
end

```
On some systems it can be a problem if a text file does not end with a newline.